greenplum-db/gpdb
# Add GUC gp_workfile_compression_overhead_limit for ZSTD buffer

### Issue

The issue is from https://github.com/greenplum-db/gpdb/issues/14785 (with more detailed analysis). In a brief, for some complex hashjoin cases with small memory, a lot of temp files (from thousands to hundreds of thousands) will be created. When compression is enabled, zstd lib creates compression buffer for each file, which might acquire mass memory. As experiment showed, a single file acquires about 1.3MB memory even without any more operation, and the total memory usage is up to 140G for about 100,000 files. So we hit various of OOM error.

### Solution

Import a new GUC `gp_workfile_comp_mem_limit` to limit the buffer used by zstd compression, and dynamically determine whether a temp file need to be compressed or not. When creating a temp file, we always check the accumulated memory usage by zstd compression. If `gp_workfile_comp_mem_limit` has been hit, the temp file (and all newer temp files) is determined not to be compressed. It means that some temp files are compressed firstly, and other temp files might not be compressed after the threshold is hit. With the approach, we will ensure the memory usage is controllable and keep the compression benefit as much as possible.

The solution also includes some code to monitor additional statistic info of temp files and output them to EXPLAIN ANALYZE.

### Test case

The relevant test case is not so easy. To repro the OOM error, we need to create 100,000 temp files or more, which is unacceptable on test env. We cannot even limit the available memory because it's beyond GPDB memory system. What I did is:

1. Use manual test.
2. Add test cases to check whether `gp_workfile_comp_mem_limit` works as expected. Once `gp_workfile_comp_mem_limit` is set to a small value, some temp files are not compressed, and we can check EXPLAIN ANALYZE to verify this.

### Manual test script

The script triggered OOM, which is eliminated by the fix. But it costs more than 16 mins on my dev env.
```
--set gp_vmem_protect_limit to a very high value

create table test_memory(a int, b text, c timestamp) distributed by (a);
insert into test_memory select id, 'test ' || id, clock_timestamp() from generate_series(1,100000000) id; 
insert into test_memory select 1,'test', now() from generate_series(1,10000) id;
insert into test_memory select id, 'test ' || id, clock_timestamp() from generate_series(1,10000000) id; 
analyze test_memory;

set statement_mem='12MB';
set gp_workfile_compression=on;
set gp_workfile_limit_files_per_query=0;

with B as (select distinct a+1 as a,b,c from test_memory)
,C as (select distinct a+2 as a,b,c from test_memory)
,D as (select a+3 as a,b,c from test_memory)
,E as (select a+4 as a,b,c from test_memory)
,F as (select (a+5)::text as a,b,c from test_memory)
select count(*) from test_memory A
inner join  B on A.a = B.a
inner join  C on A.a = C.a
inner join  D on A.a = D.a 
inner join  E on A.a = E.a 
inner join  F on A.a::text = F.a ;
```
